File: telegram/src/services/lines_service.py
"""Lines domain service wrappers."""

from __future__ import annotations

from typing import Any

from ..constants import COUNTRY_QUERY_PARAM, ERA_QUERY_PARAM, QUAFIA_QUERY_PARAM, SEA_QUERY_PARAM, TOPIC_QUERY_PARAM
from .client import ChoiceType, KatherApiClient


async def get_lines(
    client: KatherApiClient,
    *,
    poet: str | None = None,
    poem: str | None = None,
    line_type: str | None = None,
    gender: str | None = None,
    era: str | None = None,
    country: str | None = None,
    poem_type: str | None = None,
    topic: str | None = None,
    quafia: str | None = None,
    sea: str | None = None,
    page: int = 1,
    limit: int = 6,
) -> dict[str, Any]:
    """Fetch a paginated lines response matching the supplied filters."""

    return await client.get_lines(
        poet=poet,
        poem=poem,
        line_type=line_type,
        gender=gender,
        era=era,
        country=country,
        poem_type=poem_type,
        topic=topic,
        quafia=quafia,
        sea=sea,
        page=page,
        limit=limit,
    )


async def get_lines_autocomplete(
    client: KatherApiClient,
    *,
    option_name: str,
    option_value: str,
) -> list[ChoiceType]:
    """Fetch lines autocomplete choices for Telegram prompt flows."""

    normalized_option = {
        COUNTRY_QUERY_PARAM: "country",
        ERA_QUERY_PARAM: "era",
        TOPIC_QUERY_PARAM: "topic",
        QUAFIA_QUERY_PARAM: "quafia",
        SEA_QUERY_PARAM: "sea",
        "poem": "poem",
        "poet": "poet",
        "line_id": "line_id",
    }.get(option_name, option_name)
    return await client.get_lines_autocomplete(option_name=normalized_option, option_value=option_value)

File: discord/src/services/lines.service.js
import { KATHER_ROUTES, KATHER_ENDPOINTS } from '../constants/https.js';
import { COMMAND_OPTIONS } from '../constants/options.js';
import { buildQuery } from '../utils/builders/meta.builders.js';
import { fetchAndValidate, formatAutocompleteData, resolveId } from '../utils/service.helpers.js';

const LINES_AUTOCOMPLETE_URLS = {
    [COMMAND_OPTIONS.LINE.POEM]: (val) => `${KATHER_ROUTES.POEMS}?q=${encodeURIComponent(val)}`,
    [COMMAND_OPTIONS.LINE.POET]: (val) => `${KATHER_ROUTES.POETS}?q=${encodeURIComponent(val)}&limit=25`,
    [COMMAND_OPTIONS.LINE.ERA]: () => `${KATHER_ROUTES.CATALOG}${KATHER_ENDPOINTS.CATALOG.ERA}`,
    [COMMAND_OPTIONS.LINE.COUNTRY]: () => `${KATHER_ROUTES.CATALOG}${KATHER_ENDPOINTS.CATALOG.COUNTRY}`,
    [COMMAND_OPTIONS.LINE.TOPIC]: () => `${KATHER_ROUTES.CATALOG}${KATHER_ENDPOINTS.CATALOG.TOPIC}`,
    [COMMAND_OPTIONS.LINE.QUAFIA]: () => `${KATHER_ROUTES.CATALOG}${KATHER_ENDPOINTS.CATALOG.QUAFIA}`,
    [COMMAND_OPTIONS.LINE.SEA]: () => `${KATHER_ROUTES.CATALOG}${KATHER_ENDPOINTS.CATALOG.SEA}`,
};

export const getSingleLine = async ({ lineId, poemId, poetId, lineType, gender, era, country, poemType, topic, quafia, sea }) => {
    let url = `${KATHER_ROUTES.LINES}`;

    if (lineId) {
        const resolvedLineId = await resolveId(lineId, KATHER_ROUTES.LINES);
        url += `/${resolvedLineId}`
    } else if (poemId) {
        const resolvedPoemId = await resolveId(poemId, KATHER_ROUTES.POEMS);
        url += `/random${buildQuery({ poemId: resolvedPoemId })}`;
    } else if (poetId) {
        const resolvedPoetId = await resolveId(poetId, KATHER_ROUTES.POETS);
        url += `/random${buildQuery({ poetId: resolvedPoetId })}`;
    } else {
        url = `${KATHER_ROUTES.LINES}/random${buildQuery({ 
            lineType, gender, era, country, poemType, topic, quafia, sea 
        })}`;
    }

    const res = await fetchAndValidate({ url, errorMessage: 'Error fetching single line data' });
    return res.data;
};

export const getLines = async ({ poetId, poemId, lineType, gender, era, country, poemType, topic, quafia, sea, page = 1, limit = 50 }) => {
    const offset = (page - 1) * limit;
    let queryParams;

    if (poemId) {
        const resolvedPoemId = await resolveId(poemId, KATHER_ROUTES.POEMS);
        queryParams = { poemId: resolvedPoemId, limit, offset };
    } else if (poetId) {
        const resolvedPoetId = await resolveId(poetId, KATHER_ROUTES.POETS);
        queryParams = { poetId: resolvedPoetId, limit, offset };
    } else {
        queryParams = { 
            gender, era, country, poemType, topic, quafia, sea,
            lineType, limit, offset 
        };
    }

    const url = `${KATHER_ROUTES.LINES}${buildQuery(queryParams)}`;
    const res = await fetchAndValidate({ url, errorMessage: 'Error fetching lines data', expectArray: true });
    return res;
};

export const getLinesAutocomplete = async ({ optionName, optionValue, signal }) => {
    const getUrl = LINES_AUTOCOMPLETE_URLS[optionName];
    if (!getUrl) return [];

    try {
        const res = await fetchAndValidate({ url: getUrl(optionValue), signal, errorMessage: 'Lines autocomplete fetch failed' });
        return formatAutocompleteData(res.data, optionValue);
    } catch {
        return [];
    }
};

export const getSingleLineAutocomplete = async ({ optionName, optionValue, signal }) => {
    if (optionName === COMMAND_OPTIONS.LINE.LINE_ID) {
        const url = `${KATHER_ROUTES.LINES}?q=${encodeURIComponent(optionValue)}&limit=25`;
        try {
            const res = await fetchAndValidate({ url, signal, errorMessage: 'Single line autocomplete failed' });
            if (!res.data) return [];
            
            return res.data.map(ele => ({
                name: ele.content.length > 95 ? `${ele.content.substring(0, 92)}...` : ele.content,
                value: `${ele.id}`
            }));
        } catch {
            return [];
        }
    }

    return getLinesAutocomplete({ optionName, optionValue, signal });
};